### ISSUE TITLE
PPOv2Trainer & RLOOTrainer - Add Safety Check that `policy` object != `ref_policy` object

### ISSUE THREAD

**Issue by RylanSchaeffer** [huggingface/trl#2046]
### Feature request

Please throw an error or at least raise a warning if `policy` and `ref_policy` are the same object.

### Motivation

I just discovered I was making a very naive mistake. I was creating a `policy` model and passing it to `PPOv2Trainer` and `RLOOTrainer` as both `policy` and `ref_policy` since I wanted the same model to be used as the policy and the reference policy like so:

```
    ppo_trainer = PPOv2Trainer(
        config=ppo_trainer_config,
        tokenizer=policy_model_tokenizer,
        policy=policy_model,
        ref_policy=policy_model,
        reward_model=proxy_reward_model,
        value_model=value_model,
        train_dataset=train_dataset,
        eval_dataset=eval_dataset
    )
```

This was a dumb mistake that just burned quite a bit of time and compute.

I would be very grateful if someone added a sanity check to the constructors of `PPOv2Trainer` and `RLOOTrainer` that checks whether `policy` and `ref_policy` are the same object, and if so, throw an error or at least raise a warning that the programmer probably intends to use two different copies of the same language model rather than the exact same language model.

**SEPARATELY & BIZARRELY**, even if the policy model is the reference policy model, the KL divergence is non-zero and fluctuates during all of training!!!

### Your contribution

I would be happy to implement this. Just let me know if you would prefer an error or a warning.

**Comment by qgallouedec** [comment]
Good idea. I think we should implement it for all trainers. Error makes sense, as there is no way a user would want to do that.
However, handling ref_model/model is pretty tricky currently, maybe wait until #2047 is solved?

> SEPARATELY & BIZARRELY, even if the policy model is the reference policy model, the KL divergence is non-zero and fluctuates during all of training!!!

Do you think it's related to the numerical noise? Or maybe switching between eval/train?

**Comment by qgallouedec** [comment]
closed via #2057
